examples/src/main/java/kafka/examples/Producer.java

```diff
@@ -21,133 +21,159 @@
+            RecordMetadata metadata = producer.send(new ProducerRecord<>(topic, key, value)).get();
+            Utils.maybePrintRecord(numRecords, key, value, metadata);
+            return metadata;
+        } catch (AuthorizationException | UnsupportedVersionException | ProducerFencedException
+                 | FencedInstanceIdException | OutOfOrderSequenceException | SerializationException e) {
+            Utils.printErr(e.getMessage());
+            // we can't recover from these exceptions
+            shutdown();
+        } catch (KafkaException e) {
+            Utils.printErr(e.getMessage());
+        }
+        return null;
+    }
 
-    private final long startTime;
-    private final int key;
-    private final String message;
+    class ProducerCallback implements Callback {
+        private final int key;
+        private final String value;
 
-    public DemoCallBack(long startTime, int key, String message) {
-        this.startTime = startTime;
-        this.key = key;
-        this.message = message;
-    }
+        public ProducerCallback(int key, String value) {
+            this.key = key;
+            this.value = value;
+        }
 
-    /**
-     * A callback method the user can implement to provide asynchronous handling of request completion. This method will
-     * be called when the record sent to the server has been acknowledged. When exception is not null in the callback,
-     * metadata will contain the special -1 value for all fields except for topicPartition, which will be valid.
-     *
-     * @param metadata  The metadata for the record that was sent (i.e. the partition and offset). An empty metadata
-     *                  with -1 value for all fields except for topicPartition will be returned if an error occurred.
-     * @param exception The exception thrown during processing of this record. Null if no error occurred.
-     */
-    public void onCompletion(RecordMetadata metadata, Exception exception) {
-        long elapsedTime = System.currentTimeMillis() - startTime;
-        if (metadata != null) {
-            System.out.println(
-                "message(" + key + ", " + message + ") sent to partition(" + metadata.partition() +
-                    "), " +
-                    "offset(" + metadata.offset() + ") in " + elapsedTime + " ms");
-        } else {
-            exception.printStackTrace();
+        public void onCompletion(RecordMetadata metadata, Exception e) {
+            if (e != null) {
+                Utils.printErr(e.getMessage());
+                if (e instanceof AuthorizationException
+                    || e instanceof UnsupportedVersionException
+                    || e instanceof ProducerFencedException
+                    || e instanceof FencedInstanceIdException
+                    || e instanceof OutOfOrderSequenceException
+                    || e instanceof SerializationException) {
```

Comment:
I'm curious, where did you get all list exceptions that we can't recover? Are they all of them? How about `UnsupportedForMessageFormatException`? And there are also some listed [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/Callback.java#L36). Should we also add them?

Reply:
Is it safe to assume non retriable if not an instance of RetriableException? In that case we could simply do the following:

```
if (!(e instanceof RetriableException)) {
  // we can't recover from these exceptions
  shutdown();
}
```

We also need to fix the comment about OffsetMetadataTooLarge~~Exception~~.

Reply:
I think using `e instanceof RetriableException` makes sense. Had a quick look, the fatal errors are all extends ApiException, which is not a retriableException. Let's do that.